pkg/store/storepb/matcher_cache.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package storepb
+
+import (
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+const DefaultCacheSize = 200
+
+type NewItemFunc func(matcher LabelMatcher) (*labels.Matcher, error)
+
+type MatchersCache struct {
+	reg     prometheus.Registerer
+	cache   *lru.TwoQueueCache[LabelMatcher, *labels.Matcher]
+	metrics *matcherCacheMetrics
+	size    int
+}
+
+type MatcherCacheOption func(*MatchersCache)
+
+func WithPromRegistry(reg prometheus.Registerer) MatcherCacheOption {
+	return func(c *MatchersCache) {
+		c.reg = reg
+	}
+}
+
+func WithSize(size int) MatcherCacheOption {
+	return func(c *MatchersCache) {
+		c.size = size
+	}
+}
+
+func NewMatchersCache(opts ...MatcherCacheOption) (*MatchersCache, error) {
+	cache := &MatchersCache{
+		reg:  prometheus.NewRegistry(),
+		size: DefaultCacheSize,
+	}
+
+	for _, opt := range opts {
+		opt(cache)
+	}
+	cache.metrics = newMatcherCacheMetrics(cache.reg)
+
+	lruCache, err := lru.New2Q[LabelMatcher, *labels.Matcher](cache.size)
+	if err != nil {
+		return nil, err
+	}
+	cache.cache = lruCache
+
+	return cache, nil
+}
+
+func (c *MatchersCache) GetOrSet(key LabelMatcher, newItem NewItemFunc) (*labels.Matcher, error) {
+	c.metrics.requestsTotal.Inc()
+	if item, ok := c.cache.Get(key); ok {
```

Comment:
I suggest using `singleflight` here to reduce allocations even more

Reply:
Thanks! Did that.

Reply:
Kindly as from Cortex: :D

Is it possible to make this interface receive the Prometheus types, instead thanos ones, so we can reuse the same implementation on cortex?

Ex:
```
GetOrSet(t labels.MatchType, n, v string, newItem NewItemFunc) (*labels.Matcher, error)
```

Reply:
Ok..
I think that works!

But i think the interface definition should be in the storecache package? Other than that i think it would work just fine for us!